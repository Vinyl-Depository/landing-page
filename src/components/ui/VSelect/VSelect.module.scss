.container {
	position: relative;
	display: flex;
	align-items: center;
	justify-content: space-between;
	padding: 14px 10px 14px 16px;
	cursor: pointer;
	background-color: #f4f5f7;
	border-radius: 5px;

	&__text {
		font-family: Inter, sans-serif;
		font-size: 1.4rem;
		color: #474a57;
	}

	&__icon {
		width: auto;
		height: 8px;
	}

	.optionsContainer {
		position: absolute;
		bottom: 0;
		left: 0;
		display: flex;
		flex-direction: column;
		width: 100%;
		height: 350px;
		max-height: 80vh;
		overflow-y: scroll;
		background-color: #fafafa;
		border: 1px solid #18191f;
		border-radius: 5px;
		-ms-overflow-style: none;
		scrollbar-width: none;

		&::-webkit-scrollbar {
			display: none;
		}

		&__option {
			padding: 8px;
			font-size: 1.4rem;
			color: #474a57;

			&:hover {
				color: #fff;
				background-color: #afb3da;
			}
		}
	}
}
