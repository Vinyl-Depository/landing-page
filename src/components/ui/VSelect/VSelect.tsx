import React, { useState } from 'react';

import icons from 'src/icons';

import VSelectView from './VSelect.view';

interface IProps {
	readonly placeholder: string;
	readonly options: ReadonlyArray<string>;
	readonly className?: string;
	readonly iconName?: keyof typeof icons | null;
}

const VSelect: React.FC<IProps> = (props: React.PropsWithChildren<IProps>) => {
	const [selectedOptionState, setSelectedOptionState] = useState<string | null>(null);
	const [isOpenState, setIsOpenState] = useState<boolean>(false);

	const onSelectOption = (event: React.MouseEvent<HTMLSpanElement, MouseEvent>, option: string) => {
		event.stopPropagation();

		setSelectedOptionState(() => option);
		setIsOpenState(() => false);
	};

	const onCloseOptions = () => setIsOpenState(() => false);

	const onOpenOptions = () => setIsOpenState(() => true);

	return (
		<VSelectView
			placeholder={props.placeholder}
			options={props.options}
			selectedOption={selectedOptionState}
			isOpen={isOpenState}
			className={props.className}
			iconName={props.iconName}
			onSelectOption={onSelectOption}
			onCloseOptions={onCloseOptions}
			onOpenOptions={onOpenOptions}
		/>
	);
};

VSelect.displayName = 'VSelect';
VSelect.defaultProps = {};

export default React.memo(VSelect);
