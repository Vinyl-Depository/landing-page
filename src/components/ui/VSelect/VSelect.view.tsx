import React from 'react';

import { concatDiverseClasses } from '@/utils/component';

import icons from 'src/icons';

import VSvg from '../VSvg';

import classes from './VSelect.module.scss';

type UtilitiesClasses = 'container' | 'placeholder' | 'optionsContainer' | 'optionText' | 'icon';

interface IProps {
	readonly selectedIndex: number | null;
	readonly innerRef: React.RefObject<HTMLDivElement>;
	readonly placeholder: string;
	readonly options: ReadonlyArray<string>;
	readonly isOpen: boolean;
	readonly className?: string;
	readonly placeholderClassName?: string;
	readonly optionsContainerClassName?: string;
	readonly optionTextClassName?: string;
	readonly iconClassName?: string;
	readonly iconName?: keyof typeof icons | null;
	readonly onSelectOption: (event: React.MouseEvent<HTMLSpanElement, MouseEvent>, index: number) => void;
	readonly onCloseOptions: () => void;
	readonly onOpenOptions: () => void;
}

const VSelectView: React.FC<IProps> = (props: React.PropsWithChildren<IProps>) => {
	const classesMap: Record<UtilitiesClasses, string> = {
		container: concatDiverseClasses(classes['container'], props.className),
		placeholder: concatDiverseClasses(classes['container__text'], props.placeholderClassName),
		optionsContainer: concatDiverseClasses(classes['optionsContainer'], props.optionsContainerClassName),
		optionText: concatDiverseClasses(classes['optionsContainer__option'], props.optionTextClassName),
		icon: concatDiverseClasses(classes['container__icon'], props.iconClassName),
	};

	return (
		<div ref={props.innerRef} className={classesMap.container} onClick={props.onOpenOptions}>
			<span className={classesMap.placeholder}>
				{props.selectedIndex ? props.options[props.selectedIndex] : props.placeholder}
			</span>
			{props.iconName && <VSvg name={props.iconName} className={classesMap.icon} />}

			{props.isOpen && (
				<div className={classesMap.optionsContainer}>
					{props.options.map((option, index) => (
						<span
							key={index}
							className={classesMap.optionText}
							onClick={(event) => props.onSelectOption(event, index)}
						>
							{option}
						</span>
					))}
				</div>
			)}
		</div>
	);
};

VSelectView.displayName = 'VSelectView';
VSelectView.defaultProps = {};

export default React.memo(VSelectView);
