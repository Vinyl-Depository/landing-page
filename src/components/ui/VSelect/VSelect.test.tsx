import React from 'react';
import { render } from '@testing-library/react';

import VSelect from './VSelect';

describe('<VSelect>', () => {
	it('Should render the component unchanged', () => {
		const { container } = render(<VSelect placeholder="test" options={['test']} />);

		expect(container).toMatchSnapshot();
	});
});
