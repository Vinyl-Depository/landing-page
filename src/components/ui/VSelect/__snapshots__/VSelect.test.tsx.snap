// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VSelect> Should render the component unchanged 1`] = `
<div>
  <div
    class="container"
  >
    <span
      class="container__text"
    >
      test
    </span>
  </div>
</div>
`;
